ngAnimate 1.4 Refactor

matsko opened the pull request:
All of ngAnimate has been rewritten to make the internals of the
animation code more flexible, reuseable and performant.

BREAKING CHANGE: JavaSript and CSS animations can no longer be run in
parallel. With earlier versions of ngAnimate, both CSS and JS animations
would be run together when multiple animations were detected. This
feature has now been removed, however, the same effect, with even more
possibilities, can be achieved by injecting `$animateCss` into a
JavaScript-defined animation and creating custom CSS-based animations
from there. Read the ngAnimate docs for more info.

BREAKING CHANGE: The function params for `$animate.enabled()` when an
element is used are now flipped. This fix allows the function to act as
a getter when a single element param is provided.

``` js
// < 1.4
$animate.enabled(false, element);

// 1.4+
$animate.enabled(element, false);
```

BREAKING CHANGE: In addition to disabling the children of the element,
`$animate.enabled(element, false)` will now also disable animations on
the element itself.

BREAKING CHANGE: Animation-related callbacks are now fired on
`$animate.on` instead of directly being on the element.

``` js
// < 1.4
element.on('$animate:before', function(e, data) {
  if (data.event === 'enter') { ... }
});
element.off('$animate:before', fn);

// 1.4+
$animate.on(element, 'enter', function(data) {
  //...
});
$animate.off(element, 'enter', fn);
```

BREAKING CHANGE: There is no need to call `$scope.$apply` or
`$scope.$digest` inside of a animation promise callback anymore
since the promise is resolved within a digest automatically (but a
digest is not run unless the promise is chained).

``` js
// < 1.4
$animate.enter(element).then(function() {
  $scope.$apply(function() {
    $scope.explode = true;
  });
});

// 1.4+
$animate.enter(element).then(function() {
  $scope.explode = true;
});
```

BREAKING CHANGE: When an enter, leave or move animation is triggered then it
will always end any pending or active parent class based animations
(animations triggered via ngClass) in order to ensure that any CSS
styles are resolved in time.

petebacondarwin commented:
It would be helpful it was possible to see how the tests have changed from before.

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+
+function parseMaxTime(str) {
+  var maxValue = 0;
+  var values = isString(str) ?
```
`str` is always a string so this test is not needed.

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+}
+
+function getCssKeyframeDurationStyle(duration) {
+  return [ANIMATION_PROP + DURATION_KEY, duration + 's'];
```
Is it worth computing most of these constants upfront, rather than inside these functions?

* `ANIMATION_PROP + DURATION_KEY`
* `ANIMATION_PROP + DELAY_KEY`
* `TRANSITION_PROP + DELAY_KEY`
* `TRANSITION_PROP + PROPERTY_KEY`
* `ANIMATION_PROP + ANIMATION_PLAYSTATE_KEY`

matsko replied:
Fixed.

gkalpak replied:
Erm...I think the point was not only to compute the keys, but to use them too :P
I still see `ANIMATION_PROP + DURATION_KEY` (instead of `DETECT_CSS_PROPERTIES.animationDuration`).

Besides that, is this function really needed ? It is only used in one place and all it does is return an array with a pre-defined property and the passed in `duration`.

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+  return [prop, delay + 's'];
+}
+
+function blockTransitions(node, bool) {
```
could we have a more descriptive parameter name than `bool`?

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+  return [key, value];
+}
+
+function blockAnimations(node, bool) {
```
could we have a more descriptive parameter name than `bool` here too?

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+
+      if (options.transitionStyle) {
+        var transitionStyle = [TRANSITION_PROP, options.transitionStyle];
+        applyInlineStyle(node, transitionStyle);
```
It would seem to me better to just pass in the key and value separately in these cases rather than passing in an array that then needs to be unwrapped inside `applyInlineStyle`. Then we could considerably simplify that function.

matsko replied:
The reason why it's an array is so that it can be easily stored inside of `temporaryStyles`.

petebacondarwin replied:
In that case optimize the `applyInlineStyles` for this case and wrap the parameters for the other calls to it in an array. This would be faster but also make it clearer what this function expects.

petebacondarwin replied:
Ignore these comments from me here. They are not important (and probably wrong :-P )

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+  node.style[key] = value;
+}
+
+function LocalCacheLookup() {
```
Since this is really a factory function, I would rename it to `createLocalCacheLookup()` and then don't instantiate it with `new` below.

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+}
+
+function LocalCacheLookup() {
+  var cache = {};
```
use `var cache = Object.create(null)`

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+  this.$get = ['$window', '$$jqLite', '$qRaf', '$timeout', '$$rAF',
+               '$$animateRunner', '$document', '$$animateOptions', '$sniffer',
+       function($window,   $$jqLite,   $qRaf,   $timeout,   $$rAF,
+                $$animateRunner,   $document,   $$animateOptions,   $sniffer) {
```
`$qRaf` will need to go yes with your runner refactoring?

petebacondarwin replied:
By the way, I noticed that `$$jqLite` actually modifies the `JQLite` object by calling 

```js
return extend(JQLite, {
      hasClass: function(node, classes) {
        if (node.attr) node = node[0];
        return jqLiteHasClass(node, classes);
      },
      addClass: function(node, classes) {
        if (node.attr) node = node[0];
        return jqLiteAddClass(node, classes);
      },
      removeClass: function(node, classes) {
        if (node.attr) node = node[0];
        return jqLiteRemoveClass(node, classes);
      }
    });
```

This seems wrong to me.

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+
+var $AnimateCssProvider = ['$animateProvider', function($animateProvider) {
+  var gcsLookup = new LocalCacheLookup();
+  var gcsStaggerLookup = new LocalCacheLookup();
```
What does `gcs` stand for? `getComputedStyles`?

matsko replied:
Yes

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+
+    var parentCounter = 0;
+    function gcsHashFn(node, extraClasses) {
+      var KEY = "$$ngAnimateParentKey";
```
Should this constant be moved out?

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+      if (!timings) {
+        timings = computeCSSStyles($window, node, properties);
+        if (timings.animationIterationCount) {
+          timings.animationIterationCount = parseInt(timings.animationIterationCount || 1, BASE_TEN);
```
I don't think that `timings.animationIterationCount || 1` is needed as you already check for the existence of `timings.animationIterationCount`

petebacondarwin replied:
I am not sure if calling `parseInt` here is useful. The potential values for `animationIterationCount` are a floating point number (as converted already by `parseMaxTime` inside `getCSSStyles()`) or the string `"infinite"`, which `parseInt` can't handle (it returned `NaN`).

Perhaps what we should be doing is simply:

```js
if(timings.animationIterationCount === 'infinite') {
  timings.animationIterationCount = Infinity;
}
```

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+      }
+      cache[key].total++;
+      cache[key].value = value;
+    }
```
This could be optimized, since we appear to never change the value, just increment it.

```js
   put: function(key, value) {
     if (!cache[key]) {
       cache[key] = { total: 1, value: value };
     } else {
       cache[key].total++;
    }
   }
```

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+    }
+
+    function computeCachedCSSStaggerStyles(node, className, cacheKey, properties) {
+      var stagger = {};
```
Moving this object creation into an else clause below would save a wasted instantiation when there is a repeat.

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+      var isRepeated = gcsLookup.count(cacheKey) > 0;
+
+      if (isRepeated) {
+        var staggerCacheKey = cacheKey + '-stagger';
```
There is no need to generate this new key. Just use the `cacheKey` as the `gcsStaggerLookup` is independent of the `gcsLookup`

matsko replied:
Excellent catch.

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+        animationDuration:       ANIMATION_PROP  + DURATION_KEY,
+        animationDelay:          ANIMATION_PROP  + DELAY_KEY,
+        animationIterationCount: ANIMATION_PROP  + ANIMATION_ITERATION_COUNT_KEY
+      });
```
Should we make this object of constants just once are reuse?

matsko replied:
Yes good point. Maybe have it declared at the top of the file where all the other constants are defined.

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,670 @@
+      // class-based animations so that we can avoid applying CSS blocking on
+      // the element to allow for normal transitions to work (this is how enter,
+      // leave and move can perform the `ng-EVENT` methods without causing an
+      // unpected transition animation to occur).
```
unexpected

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,669 @@
+function blockTransitions(node, applyBlock) {
+  var value = applyBlock ? 'none' : '';
+  var key = TRANSITION_PROP + PROPERTY_KEY;
+  applyInlineStyle(node, key, value);
```
Change this to...

```js
applyInlineStyle(node, [key, value]);
```

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,669 @@
+function blockAnimations(node, applyBlock) {
+  var value = applyBlock ? 'paused' : '';
+  var key = ANIMATION_PROP + ANIMATION_PLAYSTATE_KEY;
+  applyInlineStyle(node, key, value);
```
Change this to...

```js
applyInlineStyle(node, [key, value]);
```

petebacondarwin commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,669 @@
+    key = key[0];
+  }
+  node.style[key] = value;
+}
```
Change this to...


```js
function applyInlineStyle(node, style) {
  node.style[style[0]] = style[1];
}
```

petebacondarwin commented on src/ng/animate.js:
```diff
@@ -1,372 +1,260 @@
+}
+
+var $$AnimateClassResolverProvider = [function() {
+  this.$get = [function() {
```
If injectables (services and providers) don't have dependencies then you don't need to wrap them in an array.

matsko replied:
Doesn't this cause a `toString()` to happen?

petebacondarwin commented on src/ng/animate.js:
```diff
@@ -1,372 +1,260 @@
+var $$AnimateClassResolverProvider = [function() {
+  this.$get = [function() {
+    return function(existing, toAdd, toRemove) {
+      var flags = {};
```
I think it is safer to use `var flags = Object.create(null);` here.

petebacondarwin replied:
Also rather than `+1`,`-1` we should have constants like: `ADD_CLASS`, `REMOVE_CLASS`.

matsko replied:
We can't use `Object.create` since `forEach` uses `hasOwnProperty` internally.

petebacondarwin commented on src/ng/animate.js:
```diff
@@ -1,372 +1,260 @@
+    var obj = {};
+    forEach(classes, function(klass) {
+      // sometimes the split leaves empty string values
+      // incase extra spaces were applied to the options
```
Perhaps instead of this you could split the `classes` parameter by a regex `/\W+/?

petebacondarwin replied:
Or even use `.replace`:

```js
function splitClasses(classes) {
  var obj = Object.create(null);
  if (isString(classes)) {
    classes.replace(/(?:^|\s)*(\S+)(?:\s|$)*/, function(klass) {
      obj[klass] = true;
    });
  }
  return obj;
});
```

petebacondarwin commented on test/ngRoute/directive/ngViewSpec.js:
```diff
@@ -837,6 +837,8 @@ describe('ngView animations', function() {
         expect($animate.queue.shift().event).toBe('enter'); //ngRepeat 3
         expect($animate.queue.shift().event).toBe('enter'); //ngRepeat 4
 
+        $$rAF.flush();
```
LGTM

petebacondarwin commented on test/ngAnimate/animationSpec.js:
```diff
@@ -0,0 +1,834 @@
+'use strict';
+
+ddescribe('$$animation', function() {
```
ddescribe!!

pkozlowski-opensource replied:
I remember we had a Grunt task for this which was failing a build in those cases...

petebacondarwin commented on test/helpers/testabilityPatch.js:
```diff
@@ -37,6 +37,13 @@ beforeEach(function() {
 afterEach(function() {
   var count, cache;
 
+  // both of these nodes are presisted across tests
```
presisted -> persisted

Does it matter if this `$$hashKey` persists across tests?

matsko replied:
It causes a collision to happen since the counter for the HashMap is reset between tests, so you may end up having the same hashCode for two different elements since the old one was stored on the HTML or BODY tag from a previous test.

petebacondarwin commented:
Do we actually need to expose the `$$animateRunner` type as an injecteable? I believe apart from returning instances of it from `$animate` methods it is private, right? This is very like [`$compile.directive.Attributes](https://docs.angularjs.org/api/ng/type/$compile.directive.Attributes), I think.

In that case we should rename it just `AnimateRunner` inside `ngAnimate` and then `$animate.AnimateRunner` for documentation purposes.

Narretz commented:
I built from the latest commits and fwiw, I am seeing obtrusive flickering on a pretty simple ng-if leave animation: http://plnkr.co/edit/qPJw2U790WUksrh9Ixxf
It's okay with current ngAnimate.
I've also seen an element not being hidden, which had the attribute `data-ng-animate="1"`set. I haven't been able to debug this futher, but it may have something to with it being in a bootstrap modal.

matsko commented:
@Narretz the flicker bug has been fixed. Thanks for pointing that out.

Not sure about the 2nd bug. Please try and see if you can reproduce it.

Narretz commented:
Thank @matsko I can confirm, first bug is fixed.
Here's a repro for the second bug:
http://plnkr.co/edit/9a4RRYnM63luz3oRg1iS?p=preview
If you open the modal, "loading" should be hidden after 400ms, but this never happens - even though the element doesn't even have animations on it.

Narretz commented:
Here's another one: http://plnkr.co/edit/7H8hg8Lq2xrAFsiqdHIo?p=preview

10 elements are repeated, and within steps of 500ms an element inside them is shown. With the refactor, the transition (height) is almost never applied. Also, when you hit the "toggle random" button, a random element is shown / hidden. Here, hiding is animated, but showing not.

petebacondarwin commented:
Great stuff @Narretz - @matsko are you able to focus on fixing these today?

matsko commented:
I'm working on these today yes. I should have the fix ready in a few hours.

matsko commented:
@Narretz what browser are you using to test these examples?

Narretz commented:
Both Firefox

matsko commented:
OK I see what's going on. I'm working on it now.

Narretz commented:
Great! :smiley_cat:

matsko commented:
Alright @Narretz  the bug has been fixed. Please double check both demos on your end.

Narretz commented:
@matsko, second bug is fixed, animations are now silky-smooth!

However, this one still exists (Chrome and FF): http://plnkr.co/edit/9a4RRYnM63luz3oRg1iS?p=preview

the element with `ng-show` is not hidden and has the class `data-ng-animate="1"` stuck on it. It also happens on Chrome, so this seems to be something different

gkalpak commented on src/ngAnimate/animateRunner.js:
```diff
@@ -0,0 +1,143 @@
+    },
+
+    pause: function() {
+      this.host.pause && this.host.pause();
```
Sorry, for coming back to this (it's just my bitterness for not "being allowed" to use such type of statements), but how about:

```js
    if (this.host.pause) this.host.pause();
    // or
    (this.host.pause || noop)();
```

matsko replied:
I used `if`.

gkalpak commented on test/ngAnimate/animateRunnerSpec.js:
```diff
@@ -0,0 +1,317 @@
+      expect(status).toBe(true);
+    }));
+
+    they("should immediately resolve if and when all runners have resolved have been $prop",
```
_have resolved have been $prop_  -->  ??

gkalpak commented on test/ngAnimate/animateRunnerSpec.js:
```diff
@@ -0,0 +1,317 @@
+
+      expect(log).toEqual([1,2,3]);
+      expect(status).toBe(true);
+    }));
```
_I have said it before, so please just ignore if you think things are fine as is._

This test does not make it clear (nor tests) what happens if the runners are completed in a different order than the order they appear in `items`.

It would it be more explicit/useful to `complete()` them in a different order (e.g. 3, 2, 1) and verify that the log order still matches their order in `items` (and not their _completed_ order) ?

matsko replied:
Saving this one for last. Thank you for the reminder.

petebacondarwin replied:
I don't think this happened...

matsko replied:
OK updated. Refactoring this test brought up an issue where I referenced `this.state` instead of `this._state`. Thanks for reminding me again about this.

gkalpak commented on src/ng/animate.js:
```diff
@@ -1,6 +1,117 @@
+  var ADD_CLASS = 1;
+  var REMOVE_CLASS = -1;
+
+  this.$get = [function() {
```
You don't need (/ shouldn't ?) use inline array notation when there is nothing to inject (as already mentioned by @petebacondarwin somewhere).
It will not call `toString()` unless the function actually has arguments.

matsko replied:
How does it know if it has no arguments?

gkalpak replied:
Functions have a `length` property that basically returns the number of arguments.
The injector tests `fn.length` and does its magic (involving `toString()`) **only** if the returned value is > 0 (aka only if the functionhas arguments).

<sub>Angular, you so smart ! Even core team members can't believe how smart you are :)</sub>

petebacondarwin replied:
I concur with @gkalpak - saves a few bytes and also speeds things up (a tiny bit) by not having to deconstruct the array.

matsko replied:
Fixed!

gkalpak commented on src/ng/animate.js:
```diff
@@ -1,6 +1,117 @@
+      toRemove = splitClasses(toRemove);
+      forEach(toRemove, function(value, key) {
+        if (flags[key] === ADD_CLASS) {
+          delete flags[key];
```
I wonder if `delete` slows things down. It doesn't seem to be necessary.
E.g. `flags[key] = flags[key] ? undefined : REMOVE_CLASS` (and then later check for a truthy value in `flags[key]`).

matsko replied:
@chirayuk confirmed. `delete` is slow. `null` is better. This has been updated.

petebacondarwin replied:
This doesn't look like it was updated.

matsko replied:
flags[key] = flags[key] === ADD_CLASS
            ? null
            : REMOVE_CLASS;

gkalpak commented on src/ng/animate.js:
```diff
@@ -8,20 +119,90 @@ var $animateMinErr = minErr('$animate');
  * Default implementation of $animate that doesn't perform any animations, instead just
- * synchronously performs DOM
- * updates and calls done() callbacks.
+ * synchronously performs DOM updates and resolved the returned runner promise.
```
_resolved_ -> _resolves_

gkalpak commented on src/ng/animate.js:
```diff
@@ -8,20 +119,90 @@ var $animateMinErr = minErr('$animate');
+ * In order to enable animations the `ngAnimate` module has to be loaded.
  *
- * To see the functional implementation check out src/ngAnimate/animate.js
+ * To see the functional implementation check out `src/ngAnimate/animate.js`.
```
Shouldn't this whole comment block be moved above `$AnimateProvider` ?

matsko replied:
Moved.

gkalpak commented on src/ng/animate.js:
```diff
@@ -1,6 +1,117 @@
+}];
+
+// this is prefixed with Core since it conflicts with
+// the animateProvider defined in ngAnimate/animate.js
```
There is no `ngAnimate/animate.js` file anymore :disappointed:

matsko replied:
Good catch. Fixed.

gkalpak commented on src/ng/animate.js:
```diff
@@ -30,33 +211,43 @@ var $AnimateProvider = ['$provide', function($provide) {
    *                           object.
    */
   this.register = function(name, factory) {
+    if (name && name.charAt(0) != '.') {
```
`!=` --> `!==` ?

gkalpak commented on src/ng/animate.js:
```diff
@@ -8,20 +119,90 @@ var $animateMinErr = minErr('$animate');
+  var postDigestElements = [];
+
+  this.$get = ['$$AnimateRunner', '$$jqLite', '$rootScope', '$$animateClassResolver',
+       function($$AnimateRunner,   $$jqLite,   $rootScope,   $$animateClassResolver) {
```
`$$jqLite` doesn't seem to be used here. Either use it (for consistency with the `ngAnimate` counterpart), or remove the dependency.

gkalpak commented on src/ng/animate.js:
```diff
@@ -81,98 +272,54 @@ var $AnimateProvider = ['$provide', function($provide) {
-        });
       }
-      return currentDefer.promise;
+      after ? after.after(element) : parent.prepend(element);
```
Big "hmm..." (style-wise).

matsko replied:
I changed `after` -> `afterElement` and `parent` => `parentElement`.

gkalpak replied:
I meant having a ternary operator for controlling execution flow.
(I know it's terse, but... :smile:)

gkalpak commented on src/ng/animate.js:
```diff
@@ -81,98 +272,54 @@ var $AnimateProvider = ['$provide', function($provide) {
-     * $animate is available in the AngularJS core, however, the ngAnimate module must be included
-     * to enable full out animation support. Otherwise, $animate will only perform simple DOM
-     * manipulation operations.
+     * It is recommended that the`$animate` service is always used when executing DOM-related procedues within directives.
```
_procedues_  -->  _procedures_

gkalpak commented on src/ng/animate.js:
```diff
@@ -81,98 +272,54 @@ var $AnimateProvider = ['$provide', function($provide) {
-     * ngAnimate module page} as well as the {@link ngAnimate.$animate ngAnimate $animate service
-     * page}.
+     * To learn more about enabling animation support, click here to visit the
+     * ngAnimate module page}.
```
`{@link ` missing.

gkalpak commented on src/ng/animate.js:
```diff
@@ -81,98 +272,54 @@ var $AnimateProvider = ['$provide', function($provide) {
-        return asyncPromise();
+      // we don't call it directly since non-existant arguments may
+      // be interpreted as null within the sub enabled function
+      on: proxy('on'),
```
I don't think `proxy`ing is necessary for these functions, since context doesn't seem to be important.
If I am not mistaken, the `proxy()` function could be removed and these calls repalces with `$$animateProvider.on/off/enabled`.

matsko replied:
OK this is good. Should I use `bind` to be safe?

gkalpak replied:
I don't know what others think, but I wouldn't use `bind` if not necessary (perfomance blah blah blah).
Tests are there for us to be safe :D

gkalpak commented on src/ng/animate.js:
```diff
@@ -180,193 +327,174 @@ var $AnimateProvider = ['$provide', function($provide) {
-       * is returned that will be resolved at a later time.
+       * @description Inserts the element into the DOM either after the `after` element (if provided) or
+       *   as the first child within the `parent` element and then triggers an animation.
+       *   A promise is returned that that will be resolved during the next digest once the animation
```
_that that_  -->  _that_

gkalpak commented on src/ng/animate.js:
```diff
@@ -180,193 +327,174 @@ var $AnimateProvider = ['$provide', function($provide) {
        * @param {DOMElement} parent the parent element which will append the element as
        *   a child (if the after element is not present)
-       * @param {DOMElement} after the sibling element which will append the element
+       * @param {DOMElement=} after the sibling element which will append the element
```
It sounds kind of weird to say that the `after` element will append the element after itself.
I would rather say _"[...] after which the element will be appended"_.

matsko replied:
I changed after, but I kept the parent element.

gkalpak commented on src/ng/animate.js:
```diff
@@ -180,193 +327,174 @@ var $AnimateProvider = ['$provide', function($provide) {
-       * @param {object=} options an optional collection of options that will be applied to the element.
+       * @description Inserts (moves) the element into its new position in the DOM either after
+       *   the `after` element (if provided) or as the first child within the `parent` element
+       *   and then triggers an animation. A promise is returned that that will be resolved
```
_that that_  -->  _that_

gkalpak commented on src/ng/animate.js:
```diff
@@ -180,193 +327,174 @@ var $AnimateProvider = ['$provide', function($provide) {
+       *
+       * @param {DOMElement} element the element which the CSS styles will be applied to
+       * @param {object} from the from (starting) CSS styles that will be applied to the element and across the animation.
+       * @param {object} to the to (starting) CSS styles that will be applied to the element and across the animation.
```
_starting_  -->  _ending_/_final_/_?_

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+ * to side-step how `$animate` and ngAnimate work, but the goal is to allow pre-existing animations or
+ * directives to create more complex animations that can be purely driven using CSS code.
+ *
+ * Note that only browsers that support CSS transitions and/or keyframe animations will only be capable of
```
_will only be capable_  ->  _will be capable_
<sub>A comma after _Note_ would also help, I think.</sub>

gkalpak commented on src/ngAnimate/animateCss.js:
````diff
@@ -0,0 +1,914 @@
+ * }]);
+ * ```
+ *
+ * Since we're adding/remove CSS classes then the CSS transition will also pick those up:
````
_remove_  -->  _removing_

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+ *
+ * @-webkit-keyframes pulse {
+ *   from { webkit-transform: scale(0.5); }
+ *   to { webkit-transform: scale(1.5); }
```
_webkit-transform_  -->  _-webkit-transform_ (2x)

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+ *
+ * `$animateCss` works in two stages: a preparation phase and an animation phase. Therefore when `$animateCss` is first called it will NOT actually
+ * start the animation. All that is going on here is that the element is being prepared for the animation (which means that the generated CSS classes are
+ * added and removed on the element). Once `$animateCss` is called it will return a an object with the following properties:
```
_a an object_  ->  _an object_

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+ * * `transition` - The raw CSS transition style that will be used (e.g. `1s linear all`).
+ * * `keyframe` - The raw CSS keyframe animation style that will be used (e.g. `1s my_animation linear`).
+ * * `from` - The starting CSS styles that will be applied at the start of the animation.
+ * * `to` - The ending CSS styles that will be applied across the animation via a CSS transition.
```
`from` and `to` are objects, right ? You might want to document it (people might try to pass in a style string).

matsko replied:
Fixed. Any idea if there is a way to create a label for these like we have in the docs for the `@param` table?

gkalpak replied:
I have no idea :confused:

Maybe you could use the (primitive) "format" used for listing action properties in `ngResource` ([docs][1], [code][2]).

[1]: https://docs.angularjs.org/api/ngResource/service/$resource#usage
[2]: https://github.com/angular/angular.js/blob/master/src/ngResource/resource.js#L124

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+ * * `to` - The ending CSS styles that will be applied across the animation via a CSS transition.
+ * * `addClass` - A space separated list of CSS classes that will be added to the element and spread across the animation.
+ * * `removeClass` - A space separated list of CSS classes that will be removed from the element and spread across the animation.
+ * * `duration` - A numeric digit representing the total duration of the transition and/or keyframe (note that a value of 1 is 1000ms). If a value of `0`
```
I suspect this doesn't have to be a _numeric digit_; it can be a number right ? E.g. 10 or 1.5 (not just **one** digit) ?

matsko replied:
Hahah good catch. Someone would've definitely gotten confused with this.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+ * * `removeClass` - A space separated list of CSS classes that will be removed from the element and spread across the animation.
+ * * `duration` - A numeric digit representing the total duration of the transition and/or keyframe (note that a value of 1 is 1000ms). If a value of `0`
+ * is provided then the animation will be skipped entirely.
+ * * `delay` - A numeric digit representing the total delay of the transition and/or keyframe (note that a value of 1 is 1000ms). If a value of `true` is
```
Same as above.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+ * CSS delay value.
+ * * `stagger` - A numeric time value representing the delay between successively animated elements
+ * ({@link ngAnimate#css-staggering-animations Click here to learn how CSS-based staggering works in ngAnimate.})
+ * * `staggerIndex` - The numeric index representing the stagger item (e.g. a value if 5 is equal to the sixth item in the stagger; therefore when a
```
_if 5_  -->  _of 5_

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+ * * `stagger` - A numeric time value representing the delay between successively animated elements
+ * ({@link ngAnimate#css-staggering-animations Click here to learn how CSS-based staggering works in ngAnimate.})
+ * * `staggerIndex` - The numeric index representing the stagger item (e.g. a value if 5 is equal to the sixth item in the stagger; therefore when a
+ * `stagger` option value if `0.6` is used then there will be a stagger delay of `600ms`)
```
I am not too familiar with staggering animations, but didn't you mean `0.1` (instead of `0.6`) ?

matsko replied:
Yes. Excellent catch.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+// http://caniuse.com/#search=transition
+if (window.ontransitionend === undefined && window.onwebkittransitionend !== undefined) {
+  CSS_PREFIX = '-webkit-';
+  TRANSITION_PROP = 'WebkitTransition';
```
Are you sure it is not `webkitTransition` (with lowercase `w`) ?

matsko replied:
This I am sure with. The code was copy/pasted from 1.3.

gkalpak replied:
I am not convinced yet...

Aha, I just found out the both `w` and `W` properties are there.
(Weird thing, Chrome DevTools' autocomplete only shows lowercase version, but both work.)

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+
+if (window.onanimationend === undefined && window.onwebkitanimationend !== undefined) {
+  CSS_PREFIX = '-webkit-';
+  ANIMATION_PROP = 'WebkitAnimation';
```
Same as above.

matsko replied:
Confirmed as the same.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+  TRANSITIONEND_EVENT = 'transitionend';
+}
+
+if (window.onanimationend === undefined && window.onwebkitanimationend !== undefined) {
```
I would prefer using `$window` for easier testing.
(I suspect it is unlikely that anyone's tests would care about this, but just in case...)

matsko replied:
Since just about all of this relies of constants I don't think we should change it. The code is also the same as 1.3.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+function computeCSSStyles($window, element, properties) {
+  var styles = Object.create(null);
+  var detectedStyles = $window.getComputedStyle(element) || {};
+  forEach(properties, function(style, prop) {
```
`style` is kind of confusing as a name. Maybe `styleKey` or something would be more clear.

matsko replied:
`style` -> `formalStyleName`

`prop` -> `actualStyleName`

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+    var val = detectedStyles[style];
+    // only numerical-based values have a digit as the first value
+    if (val) {
+      if (val.charAt(0) >= 0) {
```
This will not work correctly for negative values (e.g. -1), which are allowed on some styles (e.g. animation-delay).

matsko replied:
@gkalpak you have no idea what you have done here. I have been looking at how to do this for over a year. This changes ngAnimate forever. Thank you so much!

This is fixed now.

gkalpak replied:
Yeah, I have been watching you for over a year now, thinking "Give him some time; he'll figure it out on his own eventually", but nothing. I couldn't wait any longer, so I thought I would step up and give you a nudge in the right direction. :laughing: 

Do I get a "changed ngAnimate forever" badge or something ?

matsko replied:
I'm not being sarcastic. I had no idea about this property and now knowing this it solves one major issue with ngAnimate and provides the possibility for another. I'll confirm once all the code is in and show a demo.

gkalpak replied:
I didn't think you were sarcastic. But I doubt you have been really looking for this for over a year.
I mean, you would definitely have found it if it was important and you looked (hard).

But I am glad I happened to bring this to your attention and help you improve **my** animations :smiley:

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+  var maxValue = 0;
+  var values = str.split(/\s*,\s*/);
+  forEach(values, function(value) {
+    maxValue = Math.max(parseFloat(value) || 0, maxValue);
```
This relies on the fact that browsers convert `<time>` properties to seconds. E.g.:

    elem.style.transitionDelay = '1s, 100ms'
    =>
    <some-element style="transition-delat: 1s, 10ms;"></some-element>
    =>  but...
    getComputedStyle(elem).transitionDelay === '1s, 0.1s'

A little comment might be in place here (for maintainability's sake).

matsko replied:
// it's always safe to consider only second values and omit `ms` values since
    // getComputedStyle will always handle the conversion for us

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+  animationDuration:  ANIMATION_PROP  + DURATION_KEY
+};
+
+function computeCSSStyles($window, element, properties) {
```
_computeCSSStyles_  -->  _computeCssStyles_ (for consistency)

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+}
+
+function getCssDelayStyle(delay, isKeyframeAnimation) {
+  var prop = (isKeyframeAnimation ? ANIMATION_PROP : TRANSITION_PROP) + DELAY_KEY;
```
You can use the predefined constants to save the string concatenation.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+
+function blockTransitions(node, applyBlock) {
+  var value = applyBlock ? 'none' : '';
+  var key = TRANSITION_PROP + PROPERTY_KEY;
```
Again , you can use the predefined constant.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+  return [key, value];
+}
+
+function applyInlineStyle(node, key) {
```
`key` is probably not the most appropriate of names :smiley:

petebacondarwin replied:
`stylePair` ?

matsko replied:
styleTuple? :)

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+  var cache = Object.create(null);
+  return {
+    flush: function() {
+      cache = {};
```
Shouldn't it be `Object.create(null)` here too ?

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+      return timings;
+    }
+
+    function computeCachedCSSStaggerStyles(node, className, cacheKey, properties) {
```
_CSS_  -->  _Css_ (for consistency)

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+        //the line below will force the browser to perform a repaint so
+        //that all the animated elements within the animation frame will
+        //be properly updated and drawn on screen. This is required to
+        //ensure that the the preparation animation is properly flushed so that
```
_the the_  -->  _the_

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+      // leave and move can perform the `ng-EVENT` methods without causing an
+      // unexpected transition animation to occur).
+      var structural = method &&
+                        [' setClass ', ' addClass ',' removeClass '].indexOf(' ' + method + ' ') == -1;
```
What are the surrounding spaces for ?

matsko replied:
The spaces are a defensive way to prevent `indexOf` matching some special event value like:

```
// event = "resetClass";
['setClass', 'addClass','removeClass'].indexOf(method) == -1; //false
```

gkalpak replied:
Sorry, I don't get it. How can `indexOf()` match `setClass` with `resetClass` ?
<sub>It's called on an array, not a string.</sub>

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+      }
+
+      var setupClasses = [structuralClassName, addRemoveClassName].join(' ').trim();
+      var fullClassName =  classes + ' ' + setupClasses;
```
Extra space (just saying...)

matsko replied:
How do you find this stuff :)

gkalpak replied:
I have a thing about double-spaces. I don't find them, they find me :P

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+          // There is only one way to remove inline style properties entirely from elements.
+          // By using `removeProperty` this works, but we need to convert camel-cased CSS
+          // styles down to hyphenated values.
+          node.style.removeProperty(normalizeCssProp(entry[0]));
```
What is the problem with `node.style.someProp = ''` ?
(It seems to work on latest Chrome, Firefox, IE, unless I misunderstood the meaning of _remove inline style properties entirely_.)

matsko replied:
I remember running into an issue with IE on this. I've changed it around to not use `removeProperty` with normalization. Let's see if any browsers complain in the tests.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,914 @@
+        // will appear on screen for one animation frame causing an overbearing flicker.
+        if (options.onDone) {
+          options.onDone();
+        }
```
What happened to `options.onDone && options.onDone()` ? LOL

matsko replied:
I was a far less experienced JavaScript developer when I wrote this piece of code :P

gkalpak replied:
I see ! So you had to stick  to the style-guide.
Now you can afford improvising :smile:

petebacondarwin replied:
I think that Google Closure is actually clever enough to do this for you...

gkalpak commented on src/ngAnimate/animateCssDriver.js:
```diff
@@ -0,0 +1,204 @@
+
+    function filterCssClasses(classes) {
+      //remove all the `ng-` stuff
+      return classes.replace(/\bng-\S+\b/g, '');
```
This is not `data-` / `x-` / `ng:` friendly :disappointed:

gkalpak commented on src/ngAnimate/animateJs.js:
```diff
@@ -0,0 +1,243 @@
+          break;
+
+        default:
+          args = onDone ? [element, onDone] : [element];
```
Really guesing here, but is it OK for the `onDone` argument to be omitted (considering `options` will take its place) ?

matsko replied:
We still need to make the existing JS API work (which uses a done callback). Hopefully in 1.5 we can do something using `$animateRunner`.

gkalpak replied:
I didn't phrase it very clearly. What I meant is:

Are we OK with the fact that `args` will be either `[element, onDone, options]` **or** `[element, options]` ?
Could it lead to an errors ?
I.e. are we certain that `[element, options]` will not be applied to a function with a signature like `element, onDone, options` (thus confusing `options` for `onDone`) ?

matsko replied:
Ahh. No this was dead code from when `beforeMove` and `beforeEnter` were around. That doesn't exist anymore.

Fixed.

gkalpak commented on src/ngAnimate/animation.js:
```diff
@@ -0,0 +1,279 @@
+              anchors: []
+            };
+
+            // the anchor animations require that the from and to elements both have atleast
```
_atleast_  -->  _at least_

gkalpak commented on src/ngAnimate/animation.js:
```diff
@@ -0,0 +1,279 @@
+            };
+
+            // the anchor animations require that the from and to elements both have atleast
+            // one shared CSS class which effictively marries the two elements together to use
```
Woohoo ! The first internal Angular marriage (afaik) :cocktail: :wedding:

matsko replied:
We all get married eventually.

gkalpak commented on src/ngAnimate/animation.js:
```diff
@@ -0,0 +1,279 @@
+              matches.push(aa);
+            }
+          }
+        }
```
I don't know if the VM would optimize it anyway, but it seems redundant to assign `aa` and check it against `ng-` for every `j`:

```js
for (var i = 0; i < a.length; i++) {
  var aa = a[i];
  if (aa.substring(0,3) !== 'ng-') {
    for (var j = 0; j < b.length; j++) {
      if (aa === b[j]) {
        matches.push(aa);
      }
    }
  }
}
```

gkalpak replied:
BTW, couldn't we `break` out of the second `for-loop` after finding a match ?

matsko replied:
Optimized:

```js
        for (var i = 0; i < a.length; i++) {
          var aa = a[i];
          if (aa.substring(0,3) === 'ng-') continue;

          for (var j = 0; j < b.length; j++) {
            if (aa === b[j]) {
              matches.push(aa);
            }
          }
        }
```

gkalpak replied:
Nice ! Is there a reason not to break out of the second loop if we find a match:

```js
    ...
    for (var j = 0; j < b.length; j++) {
      if (aa === b[j]) {
        matches.push(aa);
        break;
      }
    }
    ...
```

matsko replied:
Added.

matsko commented:
@Narretz the modal bug has been fixed. Please try again.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+    // all of the remote templates being currently downloaded. If there are no
+    // templates currently downloading then the watcher will still fire anyway.
+    var deregisterWatch = $rootScope.$watch(
+      function() { return $templateRequest.totalPendingRequests; },
```
Extremely minor, but you could return `$templateRequest.totalPendingRequests === 0` to avoid calling the watch action function too often.

matsko replied:
Good idea. Fixed.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+    // $animate.on('enter', '[ng-view]', function() {
+    //
+    // });
+    // element.on('$animate:enter', j
```
`j` ?

matsko replied:
That's Igor's fault.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+          case 3:
+            callbackRegistry[event] = filterFromRegistry(entries, container, callback);
+            break;
+        }
```
Cases (2) and (3) could be replaced by:

```js
default:
  callbackRegistry[event] = filterFromRegistry(entries, container, callback);
  break;
```

matsko replied:
callbackRegistry[event] = arguments.length === 1
            ? null
            : filterFromRegistry(entries, container, callback);

gkalpak replied:
:smiley:

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+        }
+
+        function filterFromRegistry(list, matchContainer, matchCallback) {
+          var containerNode = matchContainer[0];
```
You are using `extractElementNode(container)` for registering. Shouldn't the same be used for unregistering ?

matsko replied:
Good catch.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+        var argCount = arguments.length;
+        if (isElement(bool)) {
+          var element = bool;
+          var node = element.length ? element[0] : element;
```
Could `extractElementNode()` be of any benefit here ?

matsko replied:
I changed `extractElementNode` to this:

```js
function extractElementNode(element) {
  if (!element[0]) return element;
  for (var i = 0; i < element.length; i++) {
    var elm = element[i];
    if (elm.nodeType == ELEMENT_NODE) {
      return elm;
    }
  }
}
```

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+        return queueAnimation(element, event, options);
+      },
+
+      enabled: function(bool) {
```
Seems like this function could use a 2nd argument :)

matsko replied:
It's special since the argument order changes. I added a comment

      // two signatures:
      //  (bool) - disable/enable all animations globally
      //  (element, bool) - disable/enable all animations on the element only

gkalpak replied:
Basically, I think it has 4 signatures:

```
() - global getter
(bool) - global setter
(element) - element getter
(element, bool) - element setter
```

The idea is to make the function body less complex, by showing all possible arguments and if necessary re-assign (preferrably at the top of the function and not is several places inside the body).
IIRC, the same "technique" is used in other places inside Angular (e.g. some `$http` or `$resource` callbacks).

I was thinking something like this:

```js
function (element, bool) {
  var argCount = arguments.length;

  if (argCount === 0) {
    // () - Global getter
    bool = !!animationsEnabled;    
  } else {
    var hasElement = isElement(element);
    
    if (!hasElement) {
      // (bool) - Global setter
      bool = animationsEnabled = !!element;
    } else {
      var node = element.length ? element[0] : element;
      var recordExists = disabledElementsLookup.get(node);

      if (argCount === 1) {
        // (element) - Element getter
        bool = !recordExists;
      } else {
        // (element, bool) - Element setter
        bool = !!bool;
        if (!bool) {
          disabledElementsLookup.put(node, true);
        } else if (recordExists) {
          disabledElementsLookup.remove(node);
        }
      }
    }
  }

  return bool;
}
```

I think the body of the function could be made much less complex this way.

matsko replied:
Excellent. Updated. Thank you.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+
+    function queueAnimation(element, event, options) {
+      options = $$animateOptions(element, options);
+      var node = element[0];
```
Would `extractElementNode` make sense here as well ?

gkalpak replied:
Basically, I see `element[0]` is used most of the time. I wonder if it may introduce subtle bugs.
Would it make sense to replace all occurences of `element[0]` with `extractElementNode(element)` or is it too much overhead ?

matsko replied:
It's fine to use, however here it doesn't make sense since we use `stripCommentsFromElement`. I've moved that further down so it looks like so:

      element = stripCommentsFromElement(element);
      var node = element[0];

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+
+      var isStructural = ['enter', 'move', 'leave'].indexOf(event) >= 0;
+      var existingAnimation = activeAnimationsLookup.get(node) || {};
+      var existingAnimationExists = !!existingAnimation.state;
```
<sub>I like non-existing existingAnimations !</sub>

matsko replied:
lol it took me a moment to figure out what you meant. `existingAnimationExists` -> `hasExistingAnimation`.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+      // this is a hard disable of all animations for the application or on
+      // the element itself, therefore  there is no need to continue further
+      // past this point if not enabled
+      var skipAnimations = !animationsEnabled || disabledElementsLookup.get(node);
```
Minor, but:

```js
var existingAnimation = activeAnimationsLookup.get(node) || {};
var existingAnimationExists = !!existingAnimation.state;
var skipAnimations = !animationsEnabled || disabledElementsLookup.get(node);
```

could be re-written as:

```js
var skipAnimations = !animationsEnabled || disabledElementsLookup.get(node);
var existingAnimation = (!skipAnimations && activeAnimationsLookup.get(node)) || {};
var existingAnimationExists = !!existingAnimation.state;
```

matsko replied:
Excellent!

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+          rootElementDetected = isMatchingElement(parent, $rootElement);
+        }
+
+        if (!bodyElementDetected) {
```
Maybe this could be short-circuited with `rootElementDetected && !bodyElementDetected`.
(If it is outside of the root element, we shouldn't care about body right ?)

matsko replied:
I guess that is true, but I would rather leave this as it is since it's an anti-pattern to place elements outside of body.

gkalpak replied:
Why not ?

matsko replied:
Because it will only check the body element once the rootElement is detected and when it reaches `<html>` then it's too late. We could add a special case for `parentNode.nodeType === 9` to check if it reaches the `document` object.

gkalpak replied:
You are right ! (How did I miss that ? It seems pretty obvious now...)
So, it's fine as it is, I guess. (It's not like `isMatchingElement()` is expensive or anything.)

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+
+        if (!bodyElementDetected) {
+          // we also need to ensure that the element is or will be apart of the body element
+          // otherwise it is pointless to even issue an animation to be rendered
```
<sub>(It's probably a bad practice (and possibly against some spec), but elements can exist outside of the body...)</sub>

matsko replied:
I guess that is true, but I would rather leave this as it is since it's an anti-pattern to place elements outside of body.

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,558 @@
+    }
+
+    function areAnimationsAllowed(element, parent, event) {
+      var bodyElement = jqLite($document[0].body);
```
Couldn't this be moved outside of the function (for performance reasons) ?
What are the implications ? Can the body really change ?

matsko replied:
Fixed.

gkalpak commented:
No time to look into the tests, but the (non-test) code and the whole re-organization looks absolutely fantastic.
:+1: :+1: :+1: @matsko

<sub>(Left _a couple_ of inline comments that changed `ngAnimate` for ever :P)</sub>

--
One...actualy two...more tiny things I would fix regarding the comments:
- Shorten too long comment lines (e.g. in `$animateCss`). They are hard to review.
- Add a few commas here and there

Narretz commented:
@matsko plnkr with modal works now! Awesome work!

Would it help you if you one of us prepared a commit with all typo fixes that you can cherry-pick into your branch?

petebacondarwin commented:
@Narretz I think that is an excellent idea!

matsko committed:
feat($animate): complete refactor of internal animation code

All of ngAnimate has been rewritten to make the internals of the
animation code more flexible, reuseable and performant.

BREAKING CHANGE: JavaSript and CSS animations can no longer be run in
parallel. With earlier versions of ngAnimate, both CSS and JS animations
would be run together when multiple animations were detected. This
feature has now been removed, however, the same effect, with even more
possibilities, can be achieved by injecting `$animateCss` into a
JavaScript-defined animation and creating custom CSS-based animations
from there. Read the ngAnimate docs for more info.

BREAKING CHANGE: The function params for `$animate.enabled()` when an
element is used are now flipped. This fix allows the function to act as
a getter when a single element param is provided.

```js
// < 1.4
$animate.enabled(false, element);

// 1.4+
$animate.enabled(element, false);
```

BREAKING CHANGE: In addition to disabling the children of the element,
`$animate.enabled(element, false)` will now also disable animations on
the element itself.

BREAKING CHANGE: Animation-related callbacks are now fired on
`$animate.on` instead of directly being on the element.

```js
// < 1.4
element.on('$animate:before', function(e, data) {
  if (data.event === 'enter') { ... }
});
element.off('$animate:before', fn);

// 1.4+
$animate.on(element, 'enter', function(data) {
  //...
});
$animate.off(element, 'enter', fn);
```

BREAKING CHANGE: There is no need to call `$scope.$apply` or
`$scope.$digest` inside of a animation promise callback anymore
since the promise is resolved within a digest automatically (but a
digest is not run unless the promise is chained).

```js
// < 1.4
$animate.enter(element).then(function() {
  $scope.$apply(function() {
    $scope.explode = true;
  });
});

// 1.4+
$animate.enter(element).then(function() {
  $scope.explode = true;
});
```

BREAKING CHANGE: When an enter, leave or move animation is triggered then it
will always end any pending or active parent class based animations
(animations triggered via ngClass) in order to ensure that any CSS
styles are resolved in time.

matsko commented:
@Narretz @petebacondarwin @gkalpak any typo fixes are good to go now. The PR has been updated with all the fixes that @gkalpak provided as well as the fixes that @IgorMinar suggested during our review.

`$$animateOptions` is gone now in favor of some simple private option-related helper methods.

We're almost there. Once the safari and IE-related unit tests are fixed and some extra docs are added then we're good merge.

gkalpak commented:
> BREAKING CHANGE: The function params for `$animate.enabled()` when an
element is used are now flipped. This fix allows the function to act as
a getter when a single element param is provided.

Not that it matters much (since there are several breaking changes already), but couldn't we avoid this one ? Does the order of the arguments really matters ?
<sub>(I obviously think it doesn't or else I wouldn't ask this question :))</sub>

gkalpak commented on src/ngAnimate/animateQueue.js:
```diff
@@ -0,0 +1,544 @@
+            animateChildren = value;
+          }
+        }
+
```
At this point, if `parentAnimationDetected === true` and `animateChildren === false`, we could immediately return false. There is no need to keep looping.

matsko replied:
Added. Thx.

matsko commented:
OK @petebacondarwin @IgorMinar @Narretz @gkalpak we are finally ready to merge. All green. Tested properly. Let's go!

gkalpak commented:
10, 9 . . . (6 numbers later) . . . 2, 1  ----->  :rocket: 

:sparkles: &nbsp; :clap: &nbsp; :sparkles: &nbsp; :clap: &nbsp; :sparkles:

petebacondarwin commented on modalContent.html:
```diff
@@ -0,0 +1,6 @@
+  <div class="loader" ng-show="!media.loaded">
+    Loading
+  </div>
+</div>
```
What's going on here? It doesn't seem right for this file to be in the root of the project.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,997 @@
+      var aD = timings.animationDelay;
+      var tD = timings.transitionDelay;
+      timings.maxDelay = aD && tD
+          ? Math.max(timings.animationDelay, timings.transitionDelay)
```
Minor, but you can use `aD` and `tD` here.

matsko replied:
Fixed.

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,997 @@
+    maxValue = maxValue ? Math.max(value, maxValue) : value;
+  });
+  return maxValue;
+}
```
Not sure what is the intended behaviour and if it matters, but I think the order of zero and negatie values determines the outcome (positive values are always parsed consistently).

E.g. 

```
val = '0s, -1s'  --->  maxValue === -1
val = '-1s, 0s'  --->  maxValue === 0
```

petebacondarwin replied:
From the [spec](http://dev.w3.org/csswg/css-transitions/#propdef-transition-duration)

> A negative value for transition-duration renders the declaration invalid.

matsko replied:
This is a bit tricky since `duration` and `delay` are in the same loop. This doesn't break anything, so perhaps we could just open an issue and refactor it in a couple of days. I just want this PR to go in.

gkalpak replied:
@petebacondarwin: It's the delays (e.g. [transition-delay][1]) that can have a negative value.
In any case, this is indeed unlikely to cause any problems, so it should not block this PR from landing.

[1]: http://dev.w3.org/csswg/css-transitions/#transition-delay-property

gkalpak commented on src/ngAnimate/animateCss.js:
```diff
@@ -0,0 +1,997 @@
+      var c = val.charAt(0);
+
+      // only numerical-based values have a negative sign or digit as the first value
+      if (c === '-' || c >= 0) {
```
This will miss explicitly signed positive numver, e.g. `+1s`.

matsko replied:
Fixed.

petebacondarwin replied:
According to the [spec](http://dev.w3.org/csswg/css-values-3/#numbers) this does need to be accounted for.

matsko commented:
Closed via https://github.com/angular/angular.js/commit/c8700f04fb6fb5dc21ac24de8665c0476d6db5ef
